Unable to play a fmp4 playlist.

### What version of Hls.js are you using?

v1.5.20

### What browser (including version) are you using?

Chrome

### What OS (including version) are you using?

Windows11

### Test stream

https://hlsjs.video-dev.org/demo/?src=https%3A%2F%2Fd4l93y7bh3425.cloudfront.net%2Fvod-wei%2Fhls-fmp4-test2%2Fff25dfcf-fc66-4535-8da0-efa60bf22877.m3u8&demoConfig=eyJlbmFibGVTdHJlYW1pbmciOnRydWUsImF1dG9SZWNvdmVyRXJyb3IiOnRydWUsInN0b3BPblN0YWxsIjpmYWxzZSwiZHVtcGZNUDQiOnRydWUsImxldmVsQ2FwcGluZyI6LTEsImxpbWl0TWV0cmljcyI6LTF9

### Configuration

```JavaScript
{}
```

### Additional player setup steps

_No response_

### Checklist

- [x] The issue observed is not already reported by searching on Github under https://github.com/video-dev/hls.js/issues
- [x] The issue occurs in the stable client (latest release) on https://hlsjs.video-dev.org/demo and not just on my page
- [x] The issue occurs in the latest client (main branch) on https://hlsjs-dev.video-dev.org/demo and not just on my page
- [x] The stream has correct Access-Control-Allow-Origin headers (CORS)
- [x] There are no network errors such as 404s in the browser console when trying to play the stream

### Steps to reproduce

1. Try to Play https://hlsjs.video-dev.org/demo/?src=https%3A%2F%2Fd4l93y7bh3425.cloudfront.net%2Fvod-wei%2Fhls-fmp4-test2%2Fff25dfcf-fc66-4535-8da0-efa60bf22877.m3u8&demoConfig=eyJlbmFibGVTdHJlYW1pbmciOnRydWUsImF1dG9SZWNvdmVyRXJyb3IiOnRydWUsInN0b3BPblN0YWxsIjpmYWxzZSwiZHVtcGZNUDQiOnRydWUsImxldmVsQ2FwcGluZyI6LTEsImxpbWl0TWV0cmljcyI6LTF9
2. it is giving error TypeError: Cannot read properties of undefined (reading 'toString') at toHex (https://hlsjs.video-dev.org/dist/hls.js:2063:21) at parseStsd (https://hlsjs.video-dev.org/dist/hls.js:1919:26) at


### Expected behaviour

The stream should play using hls.js

### What actually happened?

I am trying to play a stream using hls.js and it is not playing. The same stream playing fine in VLC.

### Console output

```shell
[log] > loadSource:https://d4l93y7bh3425.cloudfront.net/vod-wei/hls-fmp4-test2/ff25dfcf-fc66-4535-8da0-efa60bf22877.m3u8
stream-controller.ts:611 [log] > [stream-controller]: Trigger BUFFER_RESET
hls.ts:462 [log] > attachMedia
buffer-controller.ts:320 [log] > [buffer-controller]: created media source: MediaSource
buffer-controller.ts:1508 [log] > [buffer-controller]: Media source opened
buffer-controller.ts:1337 [log] > [buffer-controller]: checkPendingTracks (pending: 0 codec events expected: 0) {}
content_script_bundle.js:1 Attempting initialization Mon Mar 31 2025 17:49:26 GMT-0400 (Eastern Daylight Time)
level-controller.ts:370 [log] > [level-controller]: manifest loaded, 2 level(s) found, first bitrate: 387072
buffer-controller.ts:266 [log] > [buffer-controller]: 1 bufferCodec event(s) expected.
abr-controller.ts:798 [log] > [abr]: picked start tier {"videoRanges":[],"preferHDR":false,"minFramerate":0,"minBitrate":500000,"minIndex":null}
abr-controller.ts:958 [info] > [abr]: switch candidate:0->0 adjustedbw(500000)-bitrate=112928 ttfb:0.1 avgDuration:0.0 maxFetchDuration:4.0 fetchDuration:0.2 firstSelection:true codecSet: videoRange:SDR hls.loadLevel:-1
[log] > auto startLoad with configured startPosition -1
hls.ts:552 [log] > startLoad(-1)
level-controller.ts:463 [log] > [level-controller]: Switching to level 0 (SDR @387072) from level -1
level-controller.ts:668 [log] > [level-controller]: Loading level index 0 https://d4l93y7bh3425.cloudfront.net/vod-wei/hls-fmp4-test2/ff25dfcf-fc66-4535-8da0-efa60bf22877-266p-320kbps.m3u8
base-stream-controller.ts:2093 [log] > [stream-controller]: STOPPED->IDLE
base-stream-controller.ts:2093 [log] > [subtitle-stream-controller]: STOPPED->IDLE
stream-controller.ts:671 [log] > [stream-controller]: Level 0 loaded [0,8][part-8--1], cc [0, 0] duration:82.3
base-stream-controller.ts:1725 [log] > [stream-controller]: setting startPosition to 0 by default
interstitials-controller.ts:1056 [log] > [interstitials]: setSchedulePosition 0, undefined
interstitials-controller.ts:1842 [log] > [interstitials]: buffered to boundary [primary: 0.00-82.30]
interstitials-controller.ts:1294 [log] > [interstitials]: resuming [primary: 0.00-82.30]
buffer-controller.ts:1314 [log] > [buffer-controller]: Updating MediaSource duration to 82.300
base-stream-controller.ts:899 [log] > [stream-controller]: Loading main sn: initSegment of level 0 (frag:[0.000-0.000]) cc: 0 [0-8], target: 0
base-stream-controller.ts:2093 [log] > [stream-controller]: IDLE->FRAG_LOADING
[log] > [stream-controller]: FRAG_LOADING->IDLE
base-stream-controller.ts:899 [log] > [stream-controller]: Loading main sn: 0 of level 0 (frag:[0.000-10.000]) cc: 0 [0-8], target: 0
base-stream-controller.ts:2093 [log] > [stream-controller]: IDLE->FRAG_LOADING
transmuxer-interface.ts:88 [log] > injecting Web Worker for "main"
transmuxer-interface.ts:238 [log] > [transmuxer-interface]: Starting new transmux session for main sn: 0 level: 0 id: 1
        discontinuity: true
        trackSwitch: true
        contiguous: false
        accurateTimeOffset: true
        timeOffset: 0
        initSegmentChange: true
base-stream-controller.ts:512 [log] > [stream-controller]: Loaded main sn: 0 of level 0
blob:https://hlsjs.video-dev.org/86f81e1f-f38e-4cbd-9d49-ca6b3d0f2b29:1076 [log] > Debug logs enabled for "main" in hls.js version 1.6.0
transmuxer-interface.ts:420 [warn] > Error in "main" Web Worker, fallback to inline
TransmuxerInterface.onWorkerError @ transmuxer-interface.ts:420Understand this warningAI
base-stream-controller.ts:1919 [log] > [stream-controller]: Reset loading state
base-stream-controller.ts:2093 [log] > [stream-controller]: FRAG_LOADING->IDLE
base-stream-controller.ts:1919 [log] > [audio-stream-controller]: Reset loading state
main.js:745 Error event: 
{type: 'otherError', details: 'internalException', fatal: false, event: 'demuxerWorker', error: Error: Uncaught TypeError: Cannot read properties of undefined (reading 'toString')  (blob:https://…, …}
abr-controller.ts:564 [warn] > [abr]: Could not find best starting auto level. Defaulting to first in playlist 0 clamped to 0
base-stream-controller.ts:899 [log] > [stream-controller]: Loading main sn: 0 of level 0 (frag:[0.000-10.000]) cc: 0 [0-8], target: 0
base-stream-controller.ts:2093 [log] > [stream-controller]: IDLE->FRAG_LOADING
transmuxer-interface.ts:238 [log] > [transmuxer-interface]: Starting new transmux session for main sn: 0 level: 0 id: 1
        discontinuity: true
        trackSwitch: true
        contiguous: false
        accurateTimeOffset: true
        timeOffset: 0
        initSegmentChange: true
base-stream-controller.ts:2093 [log] > [stream-controller]: FRAG_LOADING->ERROR
base-stream-controller.ts:2093 [log] > [audio-stream-controller]: STOPPED->ERROR
hls.ts:575 [log] > stopLoad
base-stream-controller.ts:2093 [log] > [stream-controller]: ERROR->STOPPED
base-stream-controller.ts:2093 [log] > [audio-stream-controller]: ERROR->STOPPED
base-stream-controller.ts:2093 [log] > [subtitle-stream-controller]: IDLE->STOPPED
main.js:745 Error event: 
{type: 'otherError', details: 'internalException', err: TypeError: Cannot read properties of undefined (reading 'toString')
    at toHex (https://hlsjs.vid…, error: TypeError: Cannot read properties of undefined (reading 'toString')
    at toHex (https://hlsjs.vid…, fatal: true}
main.js:858 Fatal error : internalException
```

### Chrome media internals output

```shell

```

--- a/src/utils/mp4-tools.ts
+++ b/src/utils/mp4-tools.ts
@@ -343,11 +343,14 @@ function parseStsd(stsd: Uint8Array): StsdData {
     case 'avc4': {
       // extract profile + compatibility + level out of avcC box
       const avcCBox = findBox(sampleEntriesEnd, ['avcC'])[0];
-      codec += '.' + toHex(avcCBox[1]) + toHex(avcCBox[2]) + toHex(avcCBox[3]);
-      supplemental = parseSupplementalDoViCodec(
-        codecFourCC === 'avc1' ? 'dva1' : 'dvav',
-        sampleEntriesEnd,
-      );
+      if (avcCBox && avcCBox.length > 3) {
+        codec +=
+          '.' + toHex(avcCBox[1]) + toHex(avcCBox[2]) + toHex(avcCBox[3]);
+        supplemental = parseSupplementalDoViCodec(
+          codecFourCC === 'avc1' ? 'dva1' : 'dvav',
+          sampleEntriesEnd,
+        );
+      }
       break;
     }
     case 'mp4a': {
@@ -397,9 +400,8 @@ function parseStsd(stsd: Uint8Array): StsdData {
     }
     case 'hvc1':
     case 'hev1': {
-      const hvcCBoxes = findBox(sampleEntriesEnd, ['hvcC']);
-      if (hvcCBoxes) {
-        const hvcCBox = hvcCBoxes[0];
+      const hvcCBox = findBox(sampleEntriesEnd, ['hvcC'])[0];
+      if (hvcCBox && hvcCBox.length > 12) {
         const profileByte = hvcCBox[1];
         const profileSpace = ['', 'A', 'B', 'C'][profileByte >> 6];
         const generalProfileIdc = profileByte & 0x1f;
@@ -436,67 +438,71 @@ function parseStsd(stsd: Uint8Array): StsdData {
     }
     case 'vp09': {
       const vpcCBox = findBox(sampleEntriesEnd, ['vpcC'])[0];
-      const profile = vpcCBox[4];
-      const level = vpcCBox[5];
-      const bitDepth = (vpcCBox[6] >> 4) & 0x0f;
-      codec +=
-        '.' +
-        addLeadingZero(profile) +
-        '.' +
-        addLeadingZero(level) +
-        '.' +
-        addLeadingZero(bitDepth);
+      if (vpcCBox && vpcCBox.length > 6) {
+        const profile = vpcCBox[4];
+        const level = vpcCBox[5];
+        const bitDepth = (vpcCBox[6] >> 4) & 0x0f;
+        codec +=
+          '.' +
+          addLeadingZero(profile) +
+          '.' +
+          addLeadingZero(level) +
+          '.' +
+          addLeadingZero(bitDepth);
+      }
       break;
     }
     case 'av01': {
       const av1CBox = findBox(sampleEntriesEnd, ['av1C'])[0];
-      const profile = av1CBox[1] >>> 5;
-      const level = av1CBox[1] & 0x1f;
-      const tierFlag = av1CBox[2] >>> 7 ? 'H' : 'M';
-      const highBitDepth = (av1CBox[2] & 0x40) >> 6;
-      const twelveBit = (av1CBox[2] & 0x20) >> 5;
-      const bitDepth =
-        profile === 2 && highBitDepth
-          ? twelveBit
-            ? 12
-            : 10
-          : highBitDepth
-            ? 10
-            : 8;
-      const monochrome = (av1CBox[2] & 0x10) >> 4;
-      const chromaSubsamplingX = (av1CBox[2] & 0x08) >> 3;
-      const chromaSubsamplingY = (av1CBox[2] & 0x04) >> 2;
-      const chromaSamplePosition = av1CBox[2] & 0x03;
-      // TODO: parse color_description_present_flag
-      // default it to BT.709/limited range for now
-      // more info https://aomediacodec.github.io/av1-isobmff/#av1codecconfigurationbox-syntax
-      const colorPrimaries = 1;
-      const transferCharacteristics = 1;
-      const matrixCoefficients = 1;
-      const videoFullRangeFlag = 0;
-      codec +=
-        '.' +
-        profile +
-        '.' +
-        addLeadingZero(level) +
-        tierFlag +
-        '.' +
-        addLeadingZero(bitDepth) +
-        '.' +
-        monochrome +
-        '.' +
-        chromaSubsamplingX +
-        chromaSubsamplingY +
-        chromaSamplePosition +
-        '.' +
-        addLeadingZero(colorPrimaries) +
-        '.' +
-        addLeadingZero(transferCharacteristics) +
-        '.' +
-        addLeadingZero(matrixCoefficients) +
-        '.' +
-        videoFullRangeFlag;
-      supplemental = parseSupplementalDoViCodec('dav1', sampleEntriesEnd);
+      if (av1CBox && av1CBox.length > 2) {
+        const profile = av1CBox[1] >>> 5;
+        const level = av1CBox[1] & 0x1f;
+        const tierFlag = av1CBox[2] >>> 7 ? 'H' : 'M';
+        const highBitDepth = (av1CBox[2] & 0x40) >> 6;
+        const twelveBit = (av1CBox[2] & 0x20) >> 5;
+        const bitDepth =
+          profile === 2 && highBitDepth
+            ? twelveBit
+              ? 12
+              : 10
+            : highBitDepth
+              ? 10
+              : 8;
+        const monochrome = (av1CBox[2] & 0x10) >> 4;
+        const chromaSubsamplingX = (av1CBox[2] & 0x08) >> 3;
+        const chromaSubsamplingY = (av1CBox[2] & 0x04) >> 2;
+        const chromaSamplePosition = av1CBox[2] & 0x03;
+        // TODO: parse color_description_present_flag
+        // default it to BT.709/limited range for now
+        // more info https://aomediacodec.github.io/av1-isobmff/#av1codecconfigurationbox-syntax
+        const colorPrimaries = 1;
+        const transferCharacteristics = 1;
+        const matrixCoefficients = 1;
+        const videoFullRangeFlag = 0;
+        codec +=
+          '.' +
+          profile +
+          '.' +
+          addLeadingZero(level) +
+          tierFlag +
+          '.' +
+          addLeadingZero(bitDepth) +
+          '.' +
+          monochrome +
+          '.' +
+          chromaSubsamplingX +
+          chromaSubsamplingY +
+          chromaSamplePosition +
+          '.' +
+          addLeadingZero(colorPrimaries) +
+          '.' +
+          addLeadingZero(transferCharacteristics) +
+          '.' +
+          addLeadingZero(matrixCoefficients) +
+          '.' +
+          videoFullRangeFlag;
+        supplemental = parseSupplementalDoViCodec('dav1', sampleEntriesEnd);
+      }
       break;
     }
     case 'ac-3':

